### PR TITLE
[R] Working on Roxygen documentation

### PR DESCRIPTION
**Pull request by mayer79**
Step towards

https://github.com/dmlc/xgboost/issues/9875

After this, only a few R scripts are left for switching to the markdown style.

### CODE REVIEW

File: R-package/R/callbacks.R
Comment (mayer79) [reply R-package/R/callbacks.R:3]: Yes. The first paragraph is the title, so the second one the description, and the third the details (if not otherwise set)

File: R-package/R/callbacks.R
Comment (mayer79) [reply R-package/R/callbacks.R:417]: A square bracket is a link to a .rd file, while a [fn()] points to the rd file of the function.

File: R-package/NAMESPACE
Comment (mayer79) [reply R-package/NAMESPACE:13]: This pops up on my side since quite some months. I can easily undo this change.

Comment (trivialfis) [reply R-package/NAMESPACE:13]: I'm okay with the changes as they are generated. I am just curious about the implications. We run doc gen on the CI as well, not sure what's the cause of the change.

File: R-package/R/callbacks.R
Comment (trivialfis) [reply R-package/R/callbacks.R:3]: Thank you for the explanation.

File: R-package/R/callbacks.R
Comment (trivialfis) [reply R-package/R/callbacks.R:417]: Got it, thank you for sharing.

**Comment by david-cortes** [comment]
Some comments:
* The change of exporting S3 methods as regular functions with dots in the name would introduce issues elsewhere, particularly when dealing with namespaces directly (I can think of incorrectly dispatched functions when using `rpy2` for example). These are S3 methods and should be exported as such - not doing otherwise would be a bug.
* This will create hard-to-solve merge conflicts with other PRs like this: https://github.com/dmlc/xgboost/pull/10330
Would be ideal if these changes could be done after that PR to avoid issues. Perhaps you could base it off the last commit from that branch and then we merge them in sequence.
* In-doc links like `xgb.parameters` in `predict.xgb.Booster` now show with (empty) parentheses in the link text. I personally would prefer them without those parentheses, so that they'd match with the name of the function and the `.Rd` file. If we take the base R docs as a reference, I see there are _some_ cases where empty parentheses are added, like `coef()` in the docs for `stats::vcov`, but the parentheses are not added to the text that forms part of the link (they are plain text next to the link text).
* While it makes the files easier to inspect visually, I'm not sure it's actually an advantage to have all these docs in markdown mode:
    * We'll likely need to compose docs from different sections in the future (e.g. for parameters in `xgboost()`), and for some sections the docs are in non-markdown mode due to issues like this: https://github.com/dmlc/xgboost/issues/10329
    Would all types of docs play along if they are in different modes?
    * While the syntax is decidely more intuitive at first sight, I'm finding it extremely hard to find any information about how to do something with this new roxygen markdown style or how to address some problem. It's comparatively not as hard to look for something like "roxygen how to add named section" or "roxygen how to create link to function from different package" in a search engine, but there's hardly any results for roxygen's markdown style, and the docs from roxygen developers are not comprehensive to say the least. For example, how would one go around finding the syntax for mathematical equations for this markdown mode if we decide to add such types of docs later?

**Comment by mayer79** [comment]
@david-cortes : The file "xgb.DMatrix.R" with the unintended code block in its help file is one of the files I have not worked on yet. Thus, you should not expect it to look great. This is actually what I am trying to do with these PRs. 

As an illustration that such issues are easy to fix, see my latest commit in this PR: 

https://github.com/dmlc/xgboost/pull/10674/commits/5e57bfea9519fa2bb6e74fb2e2c77a53d43dff13

[A good and compact description of rmarkdown in Roxygen](https://roxygen2.r-lib.org/articles/markdown.html)

**Comment by trivialfis** [comment]
I will look into https://github.com/dmlc/xgboost/pull/10330 first to avoid potential conflicts.

File: R-package/NAMESPACE
Comment (terrytangyuan) [inline R-package/NAMESPACE:13]: I think this will break existing behaviors

File: R-package/NAMESPACE
Comment (terrytangyuan) [inline R-package/NAMESPACE:24]: Did we remove any explicit s3 export statements?

File: R-package/NAMESPACE
Comment (hcho3) [reply R-package/NAMESPACE:13]: This is part of the work on the new R interface: https://github.com/dmlc/xgboost/issues/9810

Comment (david-cortes) [reply R-package/NAMESPACE:13]: It shouldn't be related to the changes in other PRs. In the current master branch, triggering docs build with roxygen doesn't remove the S3 method exports.

File: R-package/NAMESPACE
Comment (mayer79) [reply R-package/NAMESPACE]: I have added a commit that fixes the problem by adding the roxygen tags `@method print xgboost` and `@method print xgb.Booster`

File: R-package/NAMESPACE
Comment (mayer79) [reply R-package/NAMESPACE]: I have now added the roxygen tags @method print xgboost and @method print xgb.Booster. This brings the NAMESPACE file back to its original state.

**Comment by trivialfis** [comment]
Hi @mayer79 , I have merged @david-cortes 's PR, could you please help rebase this one?

**Comment by terrytangyuan** [comment]
> Removing the S3 exports would indeed break things elsewhere.

Great. Glad that we caught this earlier in the PR.

**Review by trivialfis — APPROVED** [review]
Overall looks good to me, mostly reformating and reording the existing document, please let me know if there's anything specific that I should pay closer attention.

I will merge the PR if there's no further comment from @david-cortes .

**Comment by mayer79** [comment]
> Overall looks good to me, mostly reformating and reording the existing document, please let me know if there's anything specific that I should pay closer attention.
> 
> I will merge the PR if there's no further comment from @david-cortes .

After the merge, I will try to update the remaining files. Not so much is left now!

**Pull request merged**
